Wangscape/tilegen/CartesianPowerIterator.h

```diff
@@ -0,0 +1,247 @@
+#pragma once
+#include <vector>
+#include <iterator>
+#include <type_traits>
+#include <algorithm>
+#include <utility>
+#include <memory>
+#include "common.h"
+#include "CoordinatePacker.h"
+
+template<typename ForwardIt>
+class CartesianPowerIterator;
+
+namespace tilegen
+{
+template<typename ForwardIt>
+class CartesianPowerIterator : public std::iterator<std::input_iterator_tag,
+                                                    std::vector<typename ForwardIt::value_type>>
+{
+public:
+    static_assert(std::is_base_of<std::forward_iterator_tag,
+                                  typename std::iterator_traits<ForwardIt>::iterator_category>::value,
+                  "ForwardIt must be at least a ForwardIterator");
+
+    typedef std::vector<typename ForwardIt::value_type> Values;
+    typedef std::vector<ForwardIt> Iterators;
+    typedef std::vector<size_t> Coordinates;
+    typedef const Values& reference;
+    typedef const Values* pointer;
```

Comment:
`reference` and `pointer` are [required](http://en.cppreference.com/w/cpp/concept/InputIterator) by the `InputIterator` concept. I could make `ValuesRef` and `ValuesPtr`, but they would only be used in `typedef ValuesRef reference` and `typedef ValuesPtr pointer`